src/converter/one/proto/Proto.java

```diff
@@ -63,27 +66,45 @@ public Proto field(int index, byte[] bytes) {
         return this;
     }
 
+    public Proto field(int index, ByteBuffer bytes) {
+        tag(index, 2);
+        writeBytes(bytes);
+        return this;
+    }
+
     public Proto field(int index, Proto proto) {
         tag(index, 2);
         writeBytes(proto.buf, 0, proto.pos);
         return this;
     }
 
     public int startField(int index) {
+        return startField(index, DEFAULT_MAX_LEN_BYTE_COUNT).messageStart;
+    }
+
+    public Mark startField(int index, int maxLenByteCount) {
         tag(index, 2);
-        ensureCapacity(3);
-        return pos += 3;
+        ensureCapacity(maxLenByteCount);
+        pos += maxLenByteCount;
+        return new Mark(pos, maxLenByteCount);
+    }
+
+    public void commitField(int start) {
+        commitField(new Mark(start, DEFAULT_MAX_LEN_BYTE_COUNT));
     }
 
-    public void commitField(int mark) {
-        int length = pos - mark;
-        if (length >= 1 << (7 * 3)) {
+    public void commitField(Mark mark) {
+        int actualLength = pos - mark.messageStart;
+        if (actualLength >= (long) 1 << (7 * mark.maxLenByteCount)) {
             throw new IllegalArgumentException("Field too large");
         }
 
-        buf[mark - 3] = (byte) (0x80 | (length & 0x7f));
-        buf[mark - 2] = (byte) (0x80 | ((length >>> 7) & 0x7f));
-        buf[mark - 1] = (byte) (length >>> 14);
+        int lenBytesStart = mark.messageStart - mark.maxLenByteCount;
+        for (int i = 0; i < mark.maxLenByteCount - 1; ++i) {
+            buf[lenBytesStart + i] = (byte) (0x80 | (actualLength & 0x7f));
```

Comment:
Redundant `& 0x7f`.
I know this was before, but since you touch this code, it's a good opportunity to fix this.